opm/grid/GraphOfGridWrappers.cpp

```diff
@@ -0,0 +1,641 @@
+    for (const auto& v : gog)
+    {
+        gIDs[i] = v.first;
+        // lIDs are left unused
+        objWeights[i] = v.second.weight;
+        ++i;
+    }
+    *err = ZOLTAN_OK;
+}
+
+void getGraphOfGridNumEdges(void *pGraph,
+           [[maybe_unused]] int dimGlobalID,
+           [[maybe_unused]] int dimLocalID,
+                            int numCells,
+                            ZOLTAN_ID_PTR gIDs,
+           [[maybe_unused]] ZOLTAN_ID_PTR lIDs,
+                            int *numEdges,
+                            int *err)
+{
+    assert(dimGlobalID==1); // ID is a single int
+    const GraphOfGrid<Dune::CpGrid>& gog = *static_cast<const GraphOfGrid<Dune::CpGrid>*>(pGraph);
+    for (int i=0; i<numCells; ++i)
+    {
+        int nE = gog.numEdges(gIDs[i]);
+        if (nE== -1)
+        {
+            std::ostringstream ostr;
+            ostr << "getGraphOfGridNumEdges error: Vertex with ID " << gIDs[i] << " is not in graph.";
+            OpmLog::error(ostr.str());
+            *err = ZOLTAN_FATAL;
+            return;
+        }
+        numEdges[i] = nE;
+    }
+    *err = ZOLTAN_OK;
+}
+
+void getGraphOfGridEdgeList(void *pGraph,
+           [[maybe_unused]] int dimGlobalID,
+           [[maybe_unused]] int dimLocalID,
+                            int numCells,
+                            ZOLTAN_ID_PTR gIDs,
+           [[maybe_unused]] ZOLTAN_ID_PTR lIDs,
+                            int *numEdges,
+                            ZOLTAN_ID_PTR nborGIDs,
+                            int *nborProc,
+                            int weightDim,
+                            float *edgeWeights,
+                            int *err)
+{
+    assert(dimGlobalID==1); // ID is a single int
+    assert(weightDim==1); // edge weight is a single float
+    const GraphOfGrid<Dune::CpGrid>&  gog = *static_cast<const GraphOfGrid<Dune::CpGrid>*>(pGraph);
+    int id=0;
+    for (int i=0; i<numCells; ++i)
+    {
+        const auto& eList = gog.edgeList(gIDs[i]);
+        if ((int)eList.size()!=numEdges[i])
+        {
+            std::ostringstream ostr;
+            ostr << "getGraphOfGridEdgeList error: Edge number disagreement"
+                 << " between Zoltan (" << numEdges[i] << ") and Graph ("
+                 << eList.size() << ") for vertex with ID " << gIDs[i] << std::endl;
+            OpmLog::error(ostr.str());
+            *err = ZOLTAN_FATAL;
+            return;
+        }
+        for (const auto& e : eList)
+        {
+            nborGIDs[id]= e.first;
+            nborProc[id]= gog.getVertex(e.first).nproc;
+            edgeWeights[id]= e.second;
+            ++id;
+        }
+    }
+    *err = ZOLTAN_OK;
+}
+
+template<typename Zoltan_Struct>
+void setGraphOfGridZoltanGraphFunctions(Zoltan_Struct *zz,
+                      const GraphOfGrid<Dune::CpGrid>& gog,
+                                                  bool pretendNull)
+{
+    GraphOfGrid<Dune::CpGrid>* pGraph = const_cast<GraphOfGrid<Dune::CpGrid>*>(&gog);
```

Comment:
We have to pass a mutable pointer because of the Zoltan interface. Nevertheless, we know that none of the callback functions modify it and none should.
 The argument can be mutable, I simply followed the approach from ZoltanGraphFunctions.